File: android/core/src/test/kotlin/mu/core/FixtureTest.kt
package mu.core

import mu.core.CanonicalJson.render
import mu.core.CanonicalJson.toCanonical
import java.io.File
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlin.test.fail

/**
 * 契約測試：跑 contract/fixtures/cases/ 全部案例，
 * Kotlin 掃描器輸出必須與 Python 參考實作的 expected.json byte-identical。
 */
class FixtureTest {

    @Test
    fun `all fixture cases match expected json byte for byte`() {
        val casesDir = findCasesDir()
        val names = casesDir.listFiles()?.filter { it.isDirectory }?.map { it.name }?.sorted()
            ?: fail("cases dir not found")
        assertTrue(names.size >= 20, "expected >=20 cases, got ${names.size}")

        val failures = ArrayList<String>()
        for (name in names) {
            val expected = File(casesDir, "$name/expected.json").readBytes()
            val result = Scanner.scan(File(casesDir, "$name/lib"))
            val actual = render(result.toCanonical()).toByteArray(Charsets.UTF_8)
            if (!expected.contentEquals(actual)) {
                failures.add(
                    "case [$name]\n--- expected ---\n${expected.decodeToString()}\n" +
                        "--- actual ---\n${actual.decodeToString()}"
                )
            }
        }
        if (failures.isNotEmpty()) fail("${failures.size}/${names.size} cases drifted:\n\n" +
            failures.joinToString("\n\n========\n\n"))
    }

    @Test
    fun `extinf ms conversion is float-free deterministic`() {
        assertEquals(213500, Scanner.extinfToMs("213.5"))
        assertEquals(5000, Scanner.extinfToMs("5"))
        assertEquals(500, Scanner.extinfToMs(".5"))
        assertEquals(5400, Scanner.extinfToMs("5.4005"))
        assertEquals(-1500, Scanner.extinfToMs("-1.5"))
        assertEquals(null, Scanner.extinfToMs(""))
        assertEquals(null, Scanner.extinfToMs("abc"))
        assertEquals(null, Scanner.extinfToMs("1.2.3"))
    }

    @Test
    fun `canonical json escaping matches spec`() {
        val m = linkedMapOf("b" to "引", "a" to listOf<String>())
        assertEquals("{\n  \"a\": [],\n  \"b\": \"引\"\n}\n", render(m))
        assertEquals("\"\\u0001 \\n \\\" \\\\\"\n", render("\u0001 \n \" \\"))
    }

    private fun findCasesDir(): File {
        var dir = File(System.getProperty("user.dir")).absoluteFile
        while (dir != null) {
            val c = File(dir, "contract/fixtures/cases")
            if (c.isDirectory) return c
            dir = dir.parentFile
        }
        fail("contract/fixtures/cases not found from ${System.getProperty("user.dir")}")
    }
}

File: apple/MuCore/Tests/MuCoreTests/FixtureTest.swift
import XCTest
@testable import MuCore

/// 契約測試：跑 contract/fixtures/cases/ 全部案例，
/// Swift 掃描器輸出必須與 Python 參考實作的 expected.json byte-identical。
/// （ acceptance.md A4/C1 — 在 Mac 上 `swift test` ）
final class FixtureTest: XCTestCase {

    func testAllFixtureCasesMatchByteForByte() throws {
        let casesDir = try XCTUnwrap(findCasesDir(), "contract/fixtures/cases not found")
        let names = try XCTUnwrap(
            try FileManager.default.contentsOfDirectory(atPath: casesDir.path)
                .filter { name -> Bool in
                    var isDir: ObjCBool = false
                    let ok = FileManager.default.fileExists(
                        atPath: casesDir.appendingPathComponent(name).path, isDirectory: &isDir)
                    return ok && isDir.boolValue
                }
                .sorted())
        XCTAssertGreaterThanOrEqual(names.count, 20)

        var failures: [String] = []
        for name in names {
            let expected = try String(contentsOf: casesDir.appendingPathComponent("\(name)/expected.json"),
                                      encoding: .utf8)
            let result = try Scanner.scan(root: casesDir.appendingPathComponent("\(name)/lib"))
            let actual = CanonicalJson.render(CanonicalJson.canonical(result))
            if expected != actual {
                failures.append("case [\(name)]\n--- expected ---\n\(expected)\n--- actual ---\n\(actual)")
            }
        }
        if !failures.isEmpty {
            XCTFail("\(failures.count)/\(names.count) cases drifted:\n\n" +
                failures.joined(separator: "\n\n========\n\n"))
        }
    }

    func testExtinfMsConversionIsFloatFreeDeterministic() {
        XCTAssertEqual(213500, Scanner.extinfToMs("213.5"))
        XCTAssertEqual(5000, Scanner.extinfToMs("5"))
        XCTAssertEqual(500, Scanner.extinfToMs(".5"))
        XCTAssertEqual(5400, Scanner.extinfToMs("5.4005"))
        XCTAssertEqual(-1500, Scanner.extinfToMs("-1.5"))
        XCTAssertNil(Scanner.extinfToMs(""))
        XCTAssertNil(Scanner.extinfToMs("abc"))
        XCTAssertNil(Scanner.extinfToMs("1.2.3"))
    }

    func testCanonicalJsonEscapingMatchesSpec() {
        let m = CanonicalJson.JSONValue.object([
            ("b", .string("引")), ("a", .array([])),
        ])
        XCTAssertEqual("{\n  \"a\": [],\n  \"b\": \"引\"\n}\n", CanonicalJson.render(m))
        let s = CanonicalJson.render(.string("\u{01} \n \" \\"))
        XCTAssertEqual("\"\\u0001 \\n \\\" \\\\\"\n", s)
    }

    private func findCasesDir() -> URL? {
        var dir = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
        while true {
            let c = dir.appendingPathComponent("contract/fixtures/cases")
            if FileManager.default.fileExists(atPath: c.path) { return c }
            if dir.path == "/" { return nil }
            dir = dir.deletingLastPathComponent()
        }
    }
}
